### PR TITLE
[WFCORE-3999] Allow registering a custom HTTP handler for management interface

### PR DESCRIPTION
**Pull request by jmesnil**
Issue: https://issues.jboss.org/browse/WFCORE-3999

### CODE REVIEW

**Comment by jmesnil** [comment]
@rsvoboda will you be the QE contact for this one (as it relates to the Health Check feature)?

**Comment by rsvoboda** [comment]
@jmesnil yes

**Comment by rsvoboda** [comment]
Will this feature also solve current issue that the `/health` endpoint requires authentication ATM ?

**Comment by bstansberry** [comment]
This needs some clarification re: security integration.  I don't think we want the caller of this new method to take responsibility for standard security integration, but OTOH we need an option to tell the kernel to require authentication for a context, and also need to be able to have a context not require authentication.

MP Health spec says this:

```
Security
It must be possible to secure the endpoints via the usual means. The definition of 'usual means' is in this version of the specification implementation specific.

In case of a secured endpoint, accessing /metrics without valid credentials must return a 401 Unauthorized header.

A server SHOULD implement TLS encryption by default.

It is allowed to ignore security for trusted origins (e.g. localhost)
```

So, we have to be able to turn security on/off on the /health context.

MP Metrics says this:

```
Security
It must be possible to secure the endpoints via the usual means. The definition of 'usual means' is in this version of the specification implementation specific.

In case of a secured endpoint, accessing /metrics without valid credentials must return a 401 Unauthorized header.

A server SHOULD implement TLS encryption by default.

It is allowed to ignore security for trusted origins (e.g. localhost)
```

To me in our case the 'usual means' means you get the normal security configured on the management interface.  And the spec says nothing about default behavior, beyond TLS, but IMO the default should be to require authentication. Metrics can expose all sorts of information.

**Comment by rsvoboda** [comment]
For sure MP Metrics will be more tricky about the data exposure and supported security configuration

**Comment by jmesnil** [comment]
@rsvoboda what I told you previously was incorrect, access to the /health HTTP endpoint does not require authentication.

It requires that a security realm is present (and a management user must exist). If that criteria is met, it is possible to access the endpoint without client authentication.

**Comment by rsvoboda** [comment]
unzip the serve => start the server => curl http://HOST:POST/heath
This works for servers like OpenLiberty

What would be the workflow on WildFly ?

**Comment by jmesnil** [comment]
@rsvoboda current workflow is:

* unzip the server => start the server | enable management user => curl http://HOST:POST/heath

I am working with updating the proposal so that added management handler could bypass having a security realm.
I am also checking that *if* the context wants to secure access, it can do so.

**Comment by rsvoboda** [comment]
enable management user == running add-user.sh script ?

**Comment by jmesnil** [comment]
yes

**Comment by jmesnil** [comment]
I have updated the proposal. The new method has now a `requiresSecurity` boolean to specify how it integrates the management handler with the app server security

File: management/WFCORE-3999_HttpHandler_ExtensibleHttpManagement.adoc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     * Add a context for a HTTP management handler.
+     * @param contextName the name of the context. Cannot be {@code null} or empty
+     * @param requiresSecurity {@code true} if the management handler requires security.
+     * @param managementHandler HTTP management handler. Cannot be {@code null}
+     *
+     * @throws IllegalArgumentException if either parameter is invalid
+     * @throws IllegalStateException if there is already a context present named the same as {@code contextName}, either
+     *                               added via this interface or otherwise
+     */
+    void addManagementHandler(String contextName, HttpHandler managementHandler);
```
Comment (rsvoboda) [inline management/WFCORE-3999_HttpHandler_ExtensibleHttpManagement.adoc]: add `boolean requiresSecurity`

Comment (jmesnil) [reply management/WFCORE-3999_HttpHandler_ExtensibleHttpManagement.adoc]: fixed

**Comment by jmesnil** [comment]
> Does it mean add-user.sh script invocation won't be needed ? Or it will be needed, but the endpoint won't require authentication?

if security is disabled, add-user.sh script invocation is not needed

File: management/WFCORE-3999_HttpHandler_ExtensibleHttpManagement.adoc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+[code,java]
+====
+    /**
+     * Add a context for a HTTP management handler.
+     * @param contextName the name of the context. Cannot be {@code null} or empty
+     * @param requiresSecurity {@code true} if the management handler requires security.
+     * @param managementHandler HTTP management handler. Cannot be {@code null}
+     *
+     * @throws IllegalArgumentException if either parameter is invalid
+     * @throws IllegalStateException if there is already a context present named the same as {@code contextName}, either
```
Comment (rsvoboda) [inline management/WFCORE-3999_HttpHandler_ExtensibleHttpManagement.adoc]: just minor thing, this line could be a bit shorter and following a bit longer

**Pull request merged**